exchanges/subscription/list_test.go

```diff
@@ -45,3 +67,154 @@ func TestListGroupPairs(t *testing.T) {
+		{Channel: CandlesChannel, QualifiedChannel: "ohlc.futures.15m", Asset: asset.Futures, Pairs: currency.Pairs{btcusdtPair, ethusdcPair}, Interval: kline.FifteenMin},
+		{Channel: OrderbookChannel, QualifiedChannel: "book-BTCUSDT@100", Asset: asset.Spot, Pairs: currency.Pairs{btcusdtPair}, Levels: 100},
+		{Channel: OrderbookChannel, QualifiedChannel: "book-ETHUSDC@100", Asset: asset.Spot, Pairs: currency.Pairs{ethusdcPair}, Levels: 100},
+		{Channel: CandlesChannel, QualifiedChannel: "candles.spot.USDTBTC.green.15m", Asset: asset.Spot, Pairs: currency.Pairs{btcusdtPair}, Interval: kline.FifteenMin,
+			Params: map[string]any{"color": "green"}},
+		{Channel: CandlesChannel, QualifiedChannel: "candles.spot.USDCETH.red.15m", Asset: asset.Spot, Pairs: currency.Pairs{ethusdcPair}, Interval: kline.FifteenMin,
+			Params: map[string]any{"color": "green"}},
+		{Channel: CandlesChannel, QualifiedChannel: "candles.future.USDTBTC.green.15m", Asset: asset.Futures, Pairs: currency.Pairs{btcusdtPair}, Interval: kline.FifteenMin,
+			Params: map[string]any{"color": "green"}},
+		{Channel: CandlesChannel, QualifiedChannel: "candles.future.USDCETH.red.15m", Asset: asset.Futures, Pairs: currency.Pairs{ethusdcPair}, Interval: kline.FifteenMin,
+			Params: map[string]any{"color": "green"}},
+	}
+
+	if !equalLists(t, exp, got) {
+		t.FailNow() // If the first list isn't equal testing it again will duplicate test failures
+	}
+
+	got, err = l.ExpandTemplates(&mockEx{true})
+	require.NoError(t, err, "ExpandTemplates must not error")
+	exp = append(exp,
+		&Subscription{Channel: MyTradesChannel, QualifiedChannel: "trades.spot", Asset: asset.Spot, Pairs: currency.Pairs{btcusdtPair, ethusdcPair}},
+		&Subscription{Channel: MyTradesChannel, QualifiedChannel: "trades.future", Asset: asset.Futures, Pairs: currency.Pairs{btcusdtPair, ethusdcPair}},
+	)
+	equalLists(t, exp, got)
+}
+
+type mockEx struct {
+	auth bool
+}
+
+func (m *mockEx) GetEnabledPairs(_ asset.Item) (currency.Pairs, error) {
+	return currency.Pairs{btcusdtPair, ethusdcPair}, nil
+}
+
+func (m *mockEx) GetPairFormat(_ asset.Item, _ bool) (currency.PairFormat, error) {
+	return currency.PairFormat{Uppercase: true}, nil
+}
+
+func (m *mockEx) GetSubscriptionTemplateFuncs() template.FuncMap {
+	return template.FuncMap{
+		"assetName": func(a asset.Item) string {
+			if a == asset.Futures {
+				return "future"
+			}
+			return a.String()
+		},
+	}
+}
+
+func (m *mockEx) GetAssetTypes(_ bool) asset.Items            { return asset.Items{asset.Spot, asset.Futures} }
+func (m *mockEx) CanUseAuthenticatedWebsocketEndpoints() bool { return m.auth }
+
+// equalLists is a utility function to compare subscription lists and show a pretty failure message
+// It overcomes the verbose depth of assert.ElementsMatch spewConfig
+// Duplicate of internal/testing/subscriptions/equalLists
```

Comment:
Just checking we can't use an exported version?

Reply:
Not that I can work out:
* We can't export from list_test.go and use it anywhere else because we can't import test package code
* We can't use internal/testing/subscriptions here because it'd be an import cycle
* And I can't fathom any way to break that cycle, because we're using package methods

I've thought about this hard, and can't see a nice way around it yet :'(